src/Lecture4.hs

```diff
@@ -177,196 +73,76 @@ data Stats = Stats
     , statsLongest        :: MaxLen
     } deriving (Show, Eq)
 
-{-
-The 'Stats' data type has multiple fields. All these fields have
-'Semigroup' instances. This means that we can implement a 'Semigroup'
-instance for the 'Stats' type itself.
--}
-
 instance Semigroup Stats where
+    (<>)(Stats totPositions1 totSum1 absMax1 absMin1 sellMax1 sellMin1 buyMax1 buyMin1 long1)(Stats totPositions2 totSum2 absMax2 absMin2 sellMax2 sellMin2 buyMax2 buyMin2 long2)=
+	    Stats(totPositions1 <> totPositions2)
+        (totSum1 <> totSum2)
+        (absMax1 <> absMax2)
+        (absMin1 <> absMin2)
+        (sellMax1 <> sellMax2)
+        (sellMin1 <> sellMin2)
+        (buyMax1 <> buyMax2)
+        (buyMin1 <> buyMin2)
+        (long1 <> long2)
 
-
-{-
-The reason for having the 'Stats' data type is to be able to convert
-each row independently and then combine all stats into a single one.
-
-Write a function to convert a single 'Row' to 'Stats'. To implement this
-function, think about how final stats will look like if you have only a single
-row in the file.
-
-🕯 HINT: Since a single row can only be 'Buy' or 'Sell', you can't
-   populate both sell max/min and buy max/min values. In that case,
-   you can set the corresponding field to 'Nothing'.
--}
+parseRow :: String -> Maybe Row
+parseRow "" = Nothing
+parseRow rStr = if ((null $ splitRow (== ',') rStr) || ((listLen $ splitRow (== ',') rStr) /= 3)) then Nothing
+	            else
+                  do
+                     let (p:t:c:_) = splitRow (== ',') rStr
+                     t <- readMaybe t:: Maybe TradeType
+                     c <- costCheck(readMaybe c:: Maybe Int)
+                     return (Row { rowProduct = p, rowTradeType = t, rowCost = c})	
```

Comment:
You have an excellent monadic implementation 🔝 

I haven't explained the following during lectures but there's a nice way to write this parsing syntax using `do`-notation, `Monad` instance for `Maybe` and the `MonadFail` desugaring trick 🧙🏻 

You can read more about this trick here:

* https://kowainik.github.io/posts/haskell-mini-patterns#monadfail-sugar

Applying this programming pattern and a few helper functions, the `parseRow` function will look like this:

```suggestion
parseRow rStr = do
    [p, t, c] <- Just $ splitRow (== ',') rStr
    guard $ p /= ""
    t <- readMaybe t
    c <- readMaybe c
    guard $ c >= 0
    return $ Row{ rowProduct = p, rowTradeType = t, rowCost = c}
```